Debug/TimeProfiling.hpp

```diff
@@ -0,0 +1,208 @@
+/*
+ * This file is part of the source code of the software program
+ * Vampire. It is protected by applicable
+ * copyright laws.
+ *
+ * This source code is distributed under the licence found here
+ * https://vprover.github.io/license.html
+ * and in the source directory
+ */
+
+
+#ifndef __TimeProfiling__
+#define __TimeProfiling__
+
+#include "Lib/Stack.hpp"
+#include "Lib/Option.hpp"
+#include "Kernel/Ordering.hpp"
+#include "Debug/Assertion.hpp"
+#include "Kernel/Term.hpp"
+#include <chrono>
+#include "Lib/MacroUtils.hpp"
+#define ENABLE_TIME_PROFILING 1
+
+namespace Shell {
+
+#if ENABLE_TIME_PROFILING
+
+/**
+ * Enables the runtime of the current block to be measured.
+ * Time traces are stored in a tree starting with a root node that measures the runtime of the whole
+ * probram execution, and subtrees that correspond to calls to TIME_TRACE.
+ * The parameter \param name is a human-readable name to be presented when the time trace is being
+ * outputed.
+ * If you want two blocks to contribute to the same node in the tree, they have to occur in the same
+ * subtree, and have to have the same const char* name. Note that just specifying the same literal
+ * twice does not guarantee that const char* are the same, even though that is the case for some
+ * compilers. Therefore one should use one static const char* for these kind of measurements.
+ *
+ * In addition to the time profiling tree a flattened version of the tree is being outputted at the
+ * end of a program execution, that presents the sum of all the runtimes a node was encountered on
+ * any branch of the tree.
```

Comment:
It may be worth noting here that construction of these things has non-trivial overhead, so probably they should not be used in tight loops...right?

Reply:
You mean the construction of the flattened profile or the normal one? 
The flattened one is constructed in post-processing, and the other one is not too expensive (no allocations except for the first call, lookup in a `Stack` with compile-time bound (small) size, and adding up two durations). But you are right, we should deffo mention that it's not for free and this can surely make quite some impact for very short loops.

Reply:
Speaking of that. I was actually also thinking of whether it's the right thing to have this in term sharing, as I think this is a part that is called veeeery often. I only included it as there was a `TimeCounter` in the old implemenation on the same place.

Reply:
Sure, I just meant that invoking `TIME_TRACE` inside a _really_ hot loop will distort results (and performance) - I think perf is actually pretty good like you say, good job! Probably a good shout with term sharing, it's pretty hot I think.